src/v/storage/segment_index.cc

```diff
@@ -180,30 +180,69 @@ ss::future<> segment_index::truncate(model::offset o) {
     co_return co_await flush();
 }
 
+/**
+ *
+ * @return true if decoded without errors, false on a serialization error
+ *         while loading.  On all other types of error (e.g. IO), throw.
+ */
 ss::future<bool> segment_index::materialize_index() {
     auto f = co_await open();
-    auto size = co_await f.size();
-    auto buf = co_await f.dma_read_bulk<char>(0, size);
-    if (buf.empty()) {
-        co_return false;
+
+    bool decoded_ok = false;
+    std::exception_ptr ex;
+    try {
+        auto size = co_await f.size();
+        auto buf = co_await f.dma_read_bulk<char>(0, size);
+        if (buf.empty()) {
+            co_await f.close();
+            co_return false;
+        }
+        iobuf b;
+        b.append(std::move(buf));
+        try {
+            _state = serde::from_iobuf<index_state>(std::move(b));
+            decoded_ok = true;
+        } catch (const serde::serde_exception& ex) {
+            vlog(
+              stlog.info,
+              "Rebuilding index_state after decoding failure: {}",
+              ex.what());
+        }
+    } catch (...) {
+        ex = std::current_exception();
     }
-    iobuf b;
-    b.append(std::move(buf));
+
     try {
-        _state = serde::from_iobuf<index_state>(std::move(b));
-        co_return true;
-    } catch (const serde::serde_exception& ex) {
-        vlog(
-          stlog.info,
-          "Rebuilding index_state after decoding failure: {}",
-          ex.what());
-        co_return false;
+        co_await f.close();
+    } catch (...) {
+        if (!ex) {
+            // Swallow the close exception if we already had an earlier
+            // exception: the earlier one is likely to be more informative.
+            throw;
+        }
+    }
+
+    if (ex) {
+        throw ex;
+    } else {
+        co_return decoded_ok;
     }
 }
 
 ss::future<> segment_index::drop_all_data() {
     reset();
-    co_return co_await (co_await open()).truncate(0);
+    auto f = co_await open();
```

Comment:
we could introduce a helper f.e:
```c++
template<typename Func>
auto with_file(const ss::sstring& name, Func&& f) {

}
```

Reply:
It turns out seastar already has a `with_file` :partying_face: so I've converted these paths to use that.